File: app/src/main/java/com/medialink/mysqlite/adapter/NoteAdapter.java
package com.medialink.mysqlite.adapter;

import android.app.Activity;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.medialink.mysqlite.CustomOnItemClickListener;
import com.medialink.mysqlite.Note;
import com.medialink.mysqlite.NoteAddUpdateActivity;
import com.medialink.mysqlite.R;

import java.util.ArrayList;

public class NoteAdapter extends RecyclerView.Adapter<NoteAdapter.NoteViewHolder> {

    private ArrayList<Note> listNote = new ArrayList<>();
    private Activity activity;

    public NoteAdapter(Activity activity) {
        this.activity = activity;
    }

    public ArrayList<Note> getListNote() {
        return listNote;
    }

    public void setListNote(ArrayList<Note> listNote) {
        if (listNote.size() > 0) {
            this.listNote.clear();
        }
        this.listNote.addAll(listNote);
        notifyDataSetChanged();
    }

    public void addItem(Note note) {
        this.listNote.add(note);
        int position = listNote.size() - 1;
        notifyItemInserted(position);
    }

    public void updateItem(int position, Note note) {
        this.listNote.set(position, note);
        notifyItemChanged(position);
    }

    public void removeItem(int position) {
        this.listNote.remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, listNote.size());
    }

    @NonNull
    @Override
    public NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_note, parent, false);
        return new NoteViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteViewHolder holder, int position) {
        final Note note = listNote.get(position);
        holder.tvDate.setText(note.getDate());
        holder.tvTitle.setText(note.getTitle());
        holder.tvDescription.setText(note.getDescription());
        holder.cvNote.setOnClickListener(new CustomOnItemClickListener(position,
                new CustomOnItemClickListener.OnItemClickCallback() {
                    @Override
                    public void onItemClicked(View view, int position) {
                        Intent intent = new Intent(activity, NoteAddUpdateActivity.class);
                        intent.putExtra(NoteAddUpdateActivity.EXTRA_POSITION, position);
                        intent.putExtra(NoteAddUpdateActivity.EXTRA_NOTE, note);
                        activity.startActivityForResult(intent, NoteAddUpdateActivity.REQUEST_UPDATE);
                    }
                }));
    }

    @Override
    public int getItemCount() {
        return listNote.size();
    }

    public class NoteViewHolder extends RecyclerView.ViewHolder {
        private TextView tvTitle, tvDescription, tvDate;
        private CardView cvNote;

        public NoteViewHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.tv_item_title);
            tvDescription = itemView.findViewById(R.id.tv_item_description);
            tvDate = itemView.findViewById(R.id.tv_item_date);
            cvNote = itemView.findViewById(R.id.cv_item_note);
        }
    }
}

File: MySQLiteKotlin/src/main/java/com/medialink/mysqlitekotlin/adapter/NoteAdapter.kt
package com.medialink.mysqlitekotlin.adapter

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.medialink.mysqlitekotlin.CustomOnItemClickListener
import com.medialink.mysqlitekotlin.NoteAddUpdateActivity
import com.medialink.mysqlitekotlin.R
import com.medialink.mysqlitekotlin.entity.Note
import kotlinx.android.synthetic.main.item_note.view.*

class NoteAdapter(private val activity: Activity) : RecyclerView.Adapter<NoteAdapter.NoteViewHolder>() {

    var listNote = ArrayList<Note>()
        set(listNote) {
            if (listNote.size > 0) {
                this.listNote.clear()
            }
            this.listNote.addAll(listNote)
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_note, parent, false)
        return NoteViewHolder(view)
    }

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        holder.bind(listNote[position])
    }

    override fun getItemCount(): Int {
        return this.listNote.size
    }

    inner class NoteViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(note: Note) {
            with(itemView) {
                tv_item_date.text = note.date
                tv_item_title.text = note.title
                tv_item_description.text = note.description

                cv_item_note.setOnClickListener(CustomOnItemClickListener(adapterPosition,
                        object : CustomOnItemClickListener.OnItemClickCallback {
                            override fun onItemClicked(view: View, position: Int) {
                                val intent = Intent(activity, NoteAddUpdateActivity::class.java)
                                intent.putExtra(NoteAddUpdateActivity.EXTRA_POSITION, position)
                                intent.putExtra(NoteAddUpdateActivity.EXTRA_NOTE, note)
                                activity.startActivityForResult(intent, NoteAddUpdateActivity.REQUEST_UPDATE)
                            }
                        }
                ))
            }

        }
    }


    fun addItem(note: Note) {
        this.listNote.add(note)
        notifyItemInserted(this.listNote.size - 1)
    }

    fun updateItem(position: Int, note: Note) {
        this.listNote[position] = note
        notifyItemChanged(position, note)
    }

    fun removeItem(position: Int) {
        this.listNote.removeAt(position)
        notifyItemRemoved(position)
        notifyItemRangeChanged(position, this.listNote.size)
    }
}